runtime/src/snapshot_utils.rs

```diff
@@ -2814,4 +2820,211 @@ mod tests {
+            &[],
+            snapshots_dir.path(),
+            &full_snapshot_archive_path,
+            Some(&incremental_snapshot_archive_path),
+            snapshot_archive_format,
+            &genesis_config,
+            None,
+            None,
+            AccountSecondaryIndexes::default(),
+            false,
+            None,
+            AccountShrinkThreshold::default(),
+            false,
+            false,
+        )
+        .unwrap();
+        assert_eq!(
+            deserialized_bank, *bank2,
+            "Ensure rebuilding from an incremental snapshot works"
+        );
+
+        let slot = slot + 1;
+        let bank3 = Arc::new(Bank::new_from_parent(&bank2, &collector, slot));
+        // Update Account2 so that it no longer holds a reference to slot2
+        bank3
+            .transfer(lamports_to_transfer, &mint_keypair, &key2.pubkey())
+            .unwrap();
+        while !bank3.is_complete() {
+            bank3.register_tick(&Hash::new_unique());
+        }
+
+        let slot = slot + 1;
+        let bank4 = Arc::new(Bank::new_from_parent(&bank3, &collector, slot));
+        while !bank4.is_complete() {
+            bank4.register_tick(&Hash::new_unique());
+        }
+
+        // Ensure account1 has been cleaned/purged from everywhere
+        bank4.squash();
+        bank4.clean_accounts(true, false, Some(full_snapshot_slot));
+        assert_eq!(
+            bank4.get_account(&key1.pubkey()),
+            None,
+            "Ensure Account1 has been cleaned and purged from AccountsDb"
+        );
+
+        // Take an incremental snapshot and then do a roundtrip on the bank and ensure it
+        // deserializes correctly
+        let incremental_snapshot_archive_path = bank_to_incremental_snapshot_archive(
+            snapshots_dir.path(),
+            &bank4,
+            full_snapshot_slot,
+            None,
+            snapshot_archives_dir.path(),
+            snapshot_archive_format,
+            None,
+            DEFAULT_MAX_FULL_SNAPSHOT_ARCHIVES_TO_RETAIN,
+        )
+        .unwrap();
+
+        let (deserialized_bank, _) = bank_from_snapshot_archives(
+            &[accounts_dir.path().to_path_buf()],
+            &[],
+            snapshots_dir.path(),
+            &full_snapshot_archive_path,
+            Some(&incremental_snapshot_archive_path),
+            snapshot_archive_format,
+            &genesis_config,
+            None,
+            None,
+            AccountSecondaryIndexes::default(),
+            false,
+            None,
+            AccountShrinkThreshold::default(),
+            false,
+            false,
+        )
+        .unwrap();
+        assert_eq!(
+            deserialized_bank, *bank4,
+            "Ensure rebuilding from an incremental snapshot works",
+        );
+        assert_eq!(
+            deserialized_bank.get_account(&key1.pubkey()),
+            None,
+            "Ensure Account1 has not been brought back from the dead"
+        );
```

Comment:
Unfortunately, this check doesn't check that the backing index entry has been cleaned. This is because when you call `get_account()` through the bank, it funnels to this call here: https://github.com/solana-labs/solana/blob/0247c280ffb420c758142dfa6cea0af0581d0849/runtime/src/accounts.rs#L484, which filters out zero lamport accounts :)

Reply:
If this test now fails, this may be because `key1` is not found in this set here: https://github.com/solana-labs/solana/blob/0247c280ffb420c758142dfa6cea0af0581d0849/runtime/src/accounts_db.rs#L1790, so it never makes it into the zero-lamport set. 

`construct_candidate_clean_keys` gets created from `uncleaned_pubkeys` or extracted from the stores in `dirty_stores`. `uncleaned_pubkeys` gets updated on bank freeze via https://github.com/solana-labs/solana/blob/0247c280ffb420c758142dfa6cea0af0581d0849/runtime/src/accounts_db.rs#L5119, and `dirty_stores` is updated whenever an entire slot is removed from the database (dead) or we add a new root for a slot. 

Unfortunately, because the zero-lamport account will not be updated again, it won't be caught in any of the above and it doesn't make it into the `uncleaned_pubkeys` set.

We used to keep track of all the zero-lamport accounts and cleaned them on every iteration. This proved to be expensive however, hence the the comment from this PR: https://github.com/solana-labs/solana/pull/17601#discussion_r645362108, and why ultimately this was removed https://github.com/solana-labs/solana/pull/17601/files#diff-db6927be895a2b4d140c54b84392a861b4535e91afc466c5e93c243c6bf2b28aL1387.

Reply:
Maybe next to `uncleaned_pubkeys` we could introduce a `skipped_clean_zero_lamport_pubkeys` that we add to when we filter out the zero lamport update, and then consume it on the *first* clean after each full snapshot?

Reply:
I think that's because the earlier call to `bank_to_incremental_snapshot_archive()` has already removed the zero lamport account via the `clean_accounts()` call in that function. I think this was verified here: https://github.com/solana-labs/solana/pull/18870/commits/411f9da0f99fe8a685c54a455fe78751048b0179#diff-c957621ae265a12e03f7b9c5656e7aa74137223e603cc3cfcc2828dda1af5479R2964-R2973

I also added a separate clean after the bank was recreated: https://github.com/solana-labs/solana/pull/18870/commits/411f9da0f99fe8a685c54a455fe78751048b0179#diff-c957621ae265a12e03f7b9c5656e7aa74137223e603cc3cfcc2828dda1af5479R2964-R2973, this should remove the pubkey from the dirty list, and make the later clean fail, highlighting the issue described in my comment above

Reply:
Done!